pallets/gear/src/lib.rs

```diff
@@ -1374,72 +1474,23 @@ pub mod pallet {
 
-            // Check that provided `gas_limit` value does not exceed the block gas limit
-            ensure!(
-                gas_limit <= BlockGasLimitOf::<T>::get(),
-                Error::<T>::GasLimitTooHigh
-            );
-
-            let numeric_value: u128 = value.unique_saturated_into();
-            let minimum: u128 = CurrencyOf::<T>::minimum_balance().unique_saturated_into();
-
-            // Check that provided `value` equals 0 or greater than existential deposit
-            ensure!(
-                0 == numeric_value || numeric_value >= minimum,
-                Error::<T>::ValueLessThanMinimal
-            );
-
-            let schedule = T::Schedule::get();
+            Self::check_gas_limit_and_value(gas_limit, value)?;
 
-            ensure!(
-                code.len() as u32 <= schedule.limits.code_len,
-                Error::<T>::CodeTooLarge
-            );
-
-            let code = Code::try_new(code, schedule.instruction_weights.version, |module| {
-                schedule.rules(module)
-            })
-            .map_err(|e| {
-                log::debug!("Code failed to load: {:?}", e);
-                Error::<T>::FailedToConstructProgram
-            })?;
-
-            ensure!(
-                code.code().len() as u32 <= schedule.limits.code_len,
-                Error::<T>::CodeTooLarge
-            );
-
-            let code_and_id = CodeAndId::new(code);
-
-            let packet = InitPacket::new_with_gas(
+            let code_and_id = Self::check_code(code)?;
+            let packet = Self::init_packet(
+                who.clone(),
                 code_and_id.code_id(),
                 salt,
                 init_payload,
                 gas_limit,
-                value.unique_saturated_into(),
-            );
-
-            let program_id = packet.destination();
-            // Make sure there is no program with such id in program storage
-            ensure!(
-                !GearProgramPallet::<T>::program_exists(program_id),
-                Error::<T>::ProgramAlreadyExists
-            );
-
-            let reserve_fee = T::GasPrice::gas_price(gas_limit);
-
-            // First we reserve enough funds on the account to pay for `gas_limit`
-            // and to transfer declared value.
-            CurrencyOf::<T>::reserve(&who, reserve_fee + value)
-                .map_err(|_| Error::<T>::NotEnoughBalanceForReserve)?;
-
-            let origin = who.clone().into_origin();
-
-            let code_id = code_and_id.code_id();
+                value,
+            )?;
 
-            // By that call we follow the guarantee that we have in `Self::submit_code` -
+            // By that call we follow the guarantee that we have in `Self::upload_code` -
             // if there's code in storage, there's also metadata for it.
-            if let Ok(code_hash) = Self::set_code_with_metadata(code_and_id, origin) {
+            if let Ok(code_hash) =
```

Comment:
I think we need just check here manually if wasm_code exists to clarify code

Reply:
https://github.com/gear-tech/gear/pull/1241#discussion_r933784881